### PR TITLE
Some refactoring of the build system of core/

### PR DESCRIPTION
**Pull request by amadio**
This is an old branch I had from when I was working on ROOT, but at the time I could not merge it because it required at least CMake 3.12 to work, and we were still requiring only 3.9 at the time. I took the time to rework the commits against the current master yesterday and today, as rebasing was too hard. The original pull request that contained these changes was https://github.com/root-project/root/pull/4299, but the refactoring of core/ was removed. I hope now this would be ok to merge to simplify a bit how libCore is built and avoid having to add include directories by hand to targets that link against libCore.

### CODE REVIEW

File: core/rint/CMakeLists.txt
Comment (pcanal) [inline core/rint/CMakeLists.txt:25]: Does that meant that TRint (and TTabCom) are no longer part of the public interface/installed?

Comment (amadio) [reply core/rint/CMakeLists.txt:25]: I can check that, but `TRint` itself has that in its public interface, it's only private to libCore.

**Comment by pcanal** [comment]
There are change in the usage of `ROOT_OBJECT_LIBRARY` (i.e. stop using it except for Clib).  Is there a description somewhere of the advantage and/or why it is no longer needed?

**Comment by pcanal** [comment]
@phsft-bot build just on windows10/cxx14.

**Comment by pcanal** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by amadio** [comment]
@pcanal, for some targets I kept the object library to avoid compiling twice the same source files, but `Clib` indeed looks like a case that could be converted from object library to just adding sources to the `libCore` target, indeed. This is a rework of a previous branch, so I may have missed some things and could improve a bit further prior to merging.

**Comment by amadio** [comment]
@vgvassilev Is this going to cause too much trouble for the LLVM migration?

**Comment by amadio** [comment]
Rebased on master after LLVM upgrade. @vgvassilev @Axel-Naumann @bellenot Please review. Cheers.

**Comment by amadio** [comment]
Looks like one of the builds is [stuck](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/161624/console)

File: core/base/CMakeLists.txt
Comment (hageboeck) [inline core/base/CMakeLists.txt:168]: Wasn't this already added to the library in the first commit? Needed here?

**Review by hageboeck — APPROVED** [review]
Wow, bravo! That looks very good.

One question though: I think my question that's going to show up below has been answered by reading the subsequent commits (the includes removed in the beginning all moved into the respective subdirectories), but if you remove them in the first commit of this PR, ROOT doesn't configure if you hit the middle of this PR in a bisect.
Do you plan to use merge or squash?

File: core/CMakeLists.txt
Comment (amadio) [reply core/CMakeLists.txt:144]: They now come automatically from the `Core` target, added from each subdirectory where it makes sense (i.e. `core/cont/inc` added by `core/cont/CMakeLists.txt` to the `Core` target).

Comment (amadio) [reply core/CMakeLists.txt:144]: All intermediate states should work, I developed this gradually. When I removed some includes, it's because they are no longer necessary. Note that above the whole object library target is being removed, not just the includes. I am adding the source file directly to the `Core` target.

File: core/base/CMakeLists.txt
Comment (amadio) [reply core/base/CMakeLists.txt:168]: Nice catch, I think this is a "rebase artifact", this should have been in the first commit indeed. I had to rework my changes against a new version of ROOT because a normal rebase was just not possible. I can rework again to make sure it all works, but it's probably ok to just use a merge commit for the branch instead.

File: core/CMakeLists.txt
Comment (amadio) [reply core/CMakeLists.txt:144]: Actually, I forgot about the reworking on top of the new master, I will have to rebase or we should use a merge commit for this to ensure it's easy to revert if needed.

File: core/base/CMakeLists.txt
Comment (hageboeck) [reply core/base/CMakeLists.txt:168]: I say merge and be done with it, but @Axel-Naumann and @bellenot should have a chance to chime in before merging.

**Review by bellenot — APPROVED** [review]
Fine with me (I trust you), but can we wait and merge this after we branch 6.28/00 (i.e. next week)? So we don't take the risk of breaking anything in the coming release?

**Comment by amadio** [comment]
@bellenot Well, I had a milestone of 6.28 for this as I would like to see this in the release. I think the risk of breaking stuff is quite low, so this should be ok to merge before branching. It's just a build system reorganization (how libCore is built), so if libCore builds fine, everything else should be ok too. The only thing is that we should use a merge commit ("Create a merge commit" in GitHub instead of "Rebase and merge"), so that it can be easily bisected through and reverted in case that becomes necessary for any reason.

**Comment by bellenot** [comment]
@amadio "Create a merge commit" is "Not enabled for this repository"...

**Comment by amadio** [comment]
Then we have to ask Axel to merge it and push. @Axel-Naumann, can you please do that for us?

**Comment by Axel-Naumann** [comment]
After discussing with @amadio and @bellenot we will merge this right after the branch point (i.e. hopefully still on Friday). I will take care of it then.

**Comment by amadio** [comment]
@Axel-Naumann ping

**Comment by amadio** [comment]
Rebased on current master.

**Comment by amadio** [comment]
These changes passed all tests before, so I doubt the failures are related. If libCore got built and linked correctly, then this PR should be fine, as it's just a refactoring on how to build libCore.

**Comment by Axel-Naumann** [comment]
`PCH file '/mnt/build/workspace/root-pullrequests-build/build/etc//allDict.cxx.pch' not found: module file not found` is worrying. Could you check, please?

**Comment by amadio** [comment]
The context is
```
17:06:38  [10689/11146] Running utility command for SofieCompileModels_ONNX
17:06:38  error: unable to read PCH file /mnt/build/workspace/root-pullrequests-build/build/etc//allDict.cxx.pch: 'No such file or directory'
17:06:38  fatal error: PCH file '/mnt/build/workspace/root-pullrequests-build/build/etc//allDict.cxx.pch' not found: module file not found
17:06:38  FAILED: tmva/sofie/test/CMakeFiles/SofieCompileModels_ONNX.util 
```
This didn't happen before with my changes. It looks like TMVA/Sofie is trying to build a test that needs the PCH before the PCH is created (i.e. that particular test should depend on the PCH being built and it's currently not). That is, the equivalent of

https://github.com/root-project/root/blob/master/tmva/sofie/test/CMakeLists.txt#L96-L103

is missing for `SofieCompileModels_ONNX`.

**Comment by lmoneta** [comment]
Are you sure the PCH is needed for running that test ? The test is not using any dictionary or ROOT I/O but only depends on the SOFIE libraries and protobuffer. I have removed the PCH dependency since some time and I have not seen this error before. 
It is true that the target `SOFIECompileModules_ONNX` is having issue on ubunutu 18, but this is due to the old Protobuf version of the machine and I need to disable `tmva-sofie` on Ubuntu18.

**Comment by lmoneta** [comment]
This is strange, because if you look at this current nightly builds in ubuntu18 for example the PCH is not yet even built (see https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=ROOT-ubuntu18.04,SPEC=nortcxxmod,V=master/3413/console ) and the target is run and it crashes for the error in protobuf I mentioned before.

**Comment by lmoneta** [comment]
Actually I don't see this line in any builds. 
Anyway, I would suggest to merge the PR and then lets later in the nightly if we get this error again and if we need to add this dependency.

**Comment by amadio** [comment]
Rebased on latest master.

**Comment by lmoneta** [comment]
@amadio , I have tested running the executable `tmva/sofie/test/emitFromONNX` on the machine directly. 
With the current master, I don't have the PCH dependency and I have : 
```
sftnight@SFT-ubuntu-1804-2:/mnt/build/workspace/root-pullrequests-build/build/tmva/sofie/test$ rm ../../../etc/allDict.cxx.pch
sftnight@SFT-ubuntu-1804-2:/mnt/build/workspace/root-pullrequests-build/build/tmva/sofie/test$ ./emitFromONNX 
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```
with your PR I have instead: 
```
sftnight@SFT-ubuntu-1804-3:/mnt/build/workspace/root-pullrequests-build/build/tmva/sofie/test$ ./emitFromONNX 
error: unable to read PCH file /mnt/build/workspace/root-pullrequests-build/build/etc//allDict.cxx.pch: 'No such file or directory'
fatal error: PCH file '/mnt/build/workspace/root-pullrequests-build/build/etc//allDict.cxx.pch' not found: module file not found
Segmentation fault (core dumped)
```

I think before merging, it would be nice to understand why after your changes in CMAKE affecting only libCore,  a simple executable not using the ROOT I/O requires the PCH

**Comment by amadio** [comment]
Can you please try to build the PCH and try again? Also, can you run with GDB and post the stack trace? Maybe there are other missing dependencies in the test, which then tries to load them at runtime from the PCH. I wonder why there is a dependency on the PCH for the other test in any case, as I pointed out [here](https://github.com/root-project/root/blob/master/tmva/sofie/test/CMakeLists.txt#L96-L103)? Since the build is parallel, the order may slightly change after my changes, but these changes passed all tests before. I am a bit bummed that a test that is broken on master is holding off these changes after it has passed all the tests before, and has been held off twice already, for the LLVM upgrade, and the branching of 6.28, and that is not counting the first time these changes were proposed, back in 2018, when it was held for later because CMake was too old...

**Comment by lmoneta** [comment]
I am sorry for this, nothing against your changes, I would like just to understand what is happening. 
The other test uses the ROOT I/O so in that case it is expected that the PCH is needed. 
I wil try what you suggest and also on a different node (e.g. ubuntu2004) where the test works.

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/nortcxxmod

**Comment by lmoneta** [comment]
I see that the ubuntu2004 build works fine without the PCH dependency. 
See output [here](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/165830/consoleFull). There is then something strange on that old platform. Let's disable tmva-sofie there and we ignore this error. 
Please @amadio go ahead merging these changes.

**Comment by amadio** [comment]
@Axel-Naumann can you please merge? Thanks.

**Comment by Axel-Naumann** [comment]
Thanks for all your work on this, @amadio !

**Comment by amadio** [comment]
Thanks for merging!

**Comment by hahnjo** [comment]
Commit a74454ae2db9975148d9d4826d7700446b90080e broke the AddressSanitizer build - a fix is in https://github.com/root-project/root/pull/12128

**Pull request merged**
